### ISSUE TITLE
[Bug] openMSX crashing on windows-XP on certain configurations [sf#387]

### ISSUE THREAD

**Issue by openMSX-import** [openMSX/openMSX#781]
**Reported by sd-snatcher on 2010-03-03 03:44 UTC**
OpenMSX is crashing if I select some of the available devices and/or extensions.

Those I've tested so far and causes openMSX to crash on initialization, with a lot of unusual swap:

1) Machines
- Panasonic FS-A1ST

2) Extensions
- IDE
- Microsol CDX-2
- Philips NMS-1205
- Philips VY-0010

Notes:
a) The ROMs are all ok. I use the same set of ROMs on my Mac without any trouble.
b) Tested launching using Catapult and directly on the command line

Testbed details:

\- Tested with openMSX-0.7.2 and 0.7.2-11297
- Windows XP Professional SP3 (Brazilian Portuguese) with all updates
- DirectX9.0c
- Pentium-4M 1.7GHz, 1GB RAM, Nvidia GeForce 2Go 16MB VRAM, resolution=1440x1050x32 
- Renderers tested: SDL and SDLGL-FB32

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-03 03:52 UTC**
OpenMSX-0.7.2-dev11002 crash screenshot
Attached file openMSX-0.7.2-dev11002_crash.1.png:

Attachment at http://damad.be/joost/openmsx/attach/p/openmsx/bugs/_discuss/thread/3328780f/b130/attachment/openMSX-0.7.2-dev11002_crash.1.png

**Comment by openMSX-import** [comment]
**Commented by vampier on 2010-03-03 04:39 UTC**
are you doing those 4 extensions at the same time?

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-03 14:24 UTC**
No!  LOL!  :)

Sorry for the ambiguity. I wanted to mean that if I select any of those extensions (only one each time), openMSX crashes.

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-03 14:32 UTC**
I tested openMSX-0.7.2-11297 on Windows-XP under VirtualBox on Mac and the same problem happened: The Panasonic_FS-A1ST causes it to crash on initialization. Same for the IDE and the other extensions I described.

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-03 15:00 UTC**
I tested openMSX-0.6.2 on the Pentium4M machine and it worked fine. The problem seem to be specific to &gt;=0.7.2.

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2010-03-03 19:01 UTC**
So, other machines work fine?
Other extensions work fine?

What build are you using? 32 bit VC++ build?

We cannot reproduce the problem here, so we need more info...

Does it work with 0.7.0? Does it work with 0.6.3?

Is there any message when it crashes?

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-05 01:12 UTC**
Hi Manuel,

Yes, other machines worked fine. But I did make an exhaustive list. Those are some I tried:
- Panasonic FS-A1GT
- Panasonic FS-A1WSX
- Sony HB-F1XDJ
- Toshiba HX10
- Gradiente Expert 1.1
- CBIOS MSX1, MSX2, MSX2+

The same applies for the extensions. All other extensions I tried worked so far, but it's not an exhaustive list. The list includes":
- ESE Mega-SCSI
- Panasonic FS-CA1
- pac
- fmpac
- Toshiba HX-MU900
etc

I didn't tried 0.7.0 or 0.6.3. I only have 0.6.2, 0.7.2 and 0.7.2-dev11297 here.

I'm using the builds from the openmsx.sf.net page:
http://downloads.sourceforge.net/openmsx/openmsx-0.7.2-VC-x86-msi.zip
http://openmsx.sourceforge.net/temp/openmsx-debugger-latest-snapshot-windows-x86-vc-bin.zip

The 0.6.2 was laying on my notebook's HD for more than a year. I downloaded it from the openmsx.sf.net page when it was the latest release.

The only message shown when it crashes is the screenshot I attached to this bug-report .

Have you tried to run it on VirtualBox or on a Pentium4? Is the binary being compiled with optimizations that could generate instructions not available on a Pentium4?

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-05 15:36 UTC**
typo errata: Yes, other machines worked fine. But I didn't make an exhaustive list.

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2010-03-05 17:29 UTC**
I tried on VirtualBox, I don't have a Pentium 4.
But, the code should work fine on a Pentium 4...

I haven't found anyone yet who can reproduce this. (Vampier also tried.)

The only idea I have left is this: please build (or request on IRC) a debug build of the code and run with a debugger to see where it crashes exactly and to make a stack trace.

More information about this is at the utter end of http://openmsx.sourceforge.net/manual/compile.html\#contact

**Comment by openMSX-import** [comment]
**Commented by vampier on 2010-03-05 17:51 UTC**
I installed windowsXP SPK3 (MSDN ISO) on a virtual environment to test this bug. I am not able to reproduce the crash, which version are you using (what is the size of the executable) 

The only error inside openMSX I get is 'No enough memory'

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-06 00:47 UTC**
@Vampier
Those are the sha1sum and filesize of the 0.7.2-dev11002 I'm using (downloaded it from openmsx.sf.net). I already deleted the 0.7.2 (stock) and can't download it again right now because I'm traveling and conecting using my cellphone (only 250MB/month limit and max 200kbps).

sha1sum = b6a58f63c9dd658585ca2c19f80e850e5322e492  openmsx.exe
dir = 25/02/2010  21:15         6.071.296 openmsx.exe

@Manuel
I'll need that someone build the said debug version for me, but I'll not be able to use it for the next weeks because of the cell phone limits mentioned above. Also I don't have any debugger here - Do you have any free debugger to recommend?

But if it didn't gave you errors on VirtualBox, I'm beginning to suspect the problem only happens with the Brazilian Portuguese version of XP-SP3. Have you tried to test using Win-XP SP3 in Brazilian Portuguese? I believe it's ISO is available on MSDN too.

**Comment by openMSX-import** [comment]
**Commented by mfeingol on 2010-03-07 03:50 UTC**
I'll just chime in with a no-repro as well.

SD Snatcher, the best way to report a crash is to generate a dump file.  On XP, the easiest way to do this is to run ntsd.exe from a command prompt:

ntsd openmsx -machine Panasonic_FS-A1ST -ext ide

Once the process crashes, use the .dump command to generate a dump and attach the file to this bug.

I should note that to properly debug the dump, we'll need a .pdb file that matches the openmsx build you're running.  If you're running the official 0.7.2, that's easy.  Otherwise, please point us at the location where you downloaded your binary - hopefully there will be a .pdb file there too.  If not, please go back to the official 0.7.2 and re-generate the dump file.

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-12 02:45 UTC**
Ok, I've downloaded the stock openMSX-0.7.2 at a café that offered wi-fi connection.
This is the openMSX-0.7.2 I downloaded: http://downloads.sourceforge.net/openmsx/openmsx-0.7.2-VC-x86-msi.zip

I attached the dumps generated by the ntsd tool, and also included the ntsd screen output. I hope this helps.

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-12 02:47 UTC**
ntsd dumps and screen output for openMSX-0.7.2
Attached file openMSX-0.7.2_ntsd_dumps.zip:

Attachment at http://damad.be/joost/openmsx/attach/p/openmsx/bugs/_discuss/thread/3328780f/0193/attachment/openMSX-0.7.2_ntsd_dumps.zip

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2010-03-12 02:53 UTC**
All dumps were done using Windows XP Professional SP3 (Brazilian Portuguese) on the mentioned Pentium-4M. No virtual-machine was used this time.

**Comment by openMSX-import** [comment]
**Commented by mthuurne on 2010-04-06 16:24 UTC**
The crash was caused by openMSX mishandling a ZIP file in the "systemroms" directory. openMSX scans that directory for ROMs, in compressed or uncompressed form, and computes their SHA1 sum. As soon as a matching SHA1 sum is found, the scan is stopped. So it is possible openMSX crashed on a ZIP that is not needed by the machine and extensions it is started with.

The problem occurred with ZIP files that did not have the uncompressed file length filled in in the header that is before the compressed data. This field is optional and is filled in by many tools but not by all:

"This is immediately followed by the compressed data. If bit 3 of the 
general-purpose flags field is set, then the CRC-32 and file sizes are not 
known when the header is written, and they are appended (in a 12-byte 
structure) immediately after the compressed data. The fields in the local 
header are filled with zero."

http://en.wikipedia.org/wiki/ZIP\_\(file\_format\)\#The\_structure\_of\_a\_ZIP\_file

Inside openMSX, ZlibInflate::inflate() got 0 as the size hint for the decompressed data. This caused it to allocate a buffer with an initial size of 0. Then it tried to get a pointer to the first byte in that buffer, which is illegal. When compiled with GNU's STL, that returns a NULL pointer, which is handled gracefully by zlib. When compiled with VC++, it triggers an assert (debug mode) or an invalid pointer (release mode), resulting in a crash.

I fixed it in r11385, by adding a minimum initial buffer size. To avoid other kinds of trouble with weird ZIP files, I also added a maximum initial buffer size.

Thanks to FRS, Max and Patrick for helping in finding out what was going on here.
